quickwit/quickwit-ingest/src/ingest_v2/state.rs

```diff
@@ -0,0 +1,212 @@
+use std::sync::{Arc, Weak};
+
+use fnv::FnvHashMap;
+use mrecordlog::error::{DeleteQueueError, TruncateError};
+use mrecordlog::MultiRecordLog;
+use quickwit_common::rate_limiter::RateLimiter;
+use quickwit_proto::ingest::ingester::{IngesterStatus, ObservationMessage};
+use quickwit_proto::ingest::IngestV2Result;
+use quickwit_proto::types::{Position, QueueId};
+use tokio::sync::{watch, Mutex, MutexGuard, RwLock, RwLockWriteGuard};
+use tracing::{error, info};
+
+use super::models::IngesterShard;
+use super::rate_meter::RateMeter;
+use super::replication::{ReplicationStreamTaskHandle, ReplicationTaskHandle};
+use crate::{FollowerId, LeaderId};
+
+/// Stores the state of the ingester and attempts to prevent deadlocks by exposing an API that
+/// guarantees that the internal data structures are always locked in the same order: first,
+/// `inner`, then `mrecordlog`.
+///
+/// `lock_partially` locks `inner` only, while `lock_fully` locks both `inner` and `mrecordlog`. Use
+/// the former when you only need to access the in-memory state of the ingester and the latter when
+/// you need to access both the in-memory state AND the WAL.
+#[derive(Clone)]
+pub(super) struct IngesterState {
+    // `inner` is a mutex because it's almost always accessed mutably.
+    inner: Arc<Mutex<InnerIngesterState>>,
+    mrecordlog: Arc<RwLock<MultiRecordLog>>,
+}
+
+pub(super) struct InnerIngesterState {
+    pub shards: FnvHashMap<QueueId, IngesterShard>,
+    pub rate_trackers: FnvHashMap<QueueId, (RateLimiter, RateMeter)>,
+    // Replication stream opened with followers.
+    pub replication_streams: FnvHashMap<FollowerId, ReplicationStreamTaskHandle>,
+    // Replication tasks running for each replication stream opened with leaders.
+    pub replication_tasks: FnvHashMap<LeaderId, ReplicationTaskHandle>,
+    pub status: IngesterStatus,
+    pub observation_tx: watch::Sender<IngestV2Result<ObservationMessage>>,
+}
+
+impl IngesterState {
+    pub fn new(
+        mrecordlog: MultiRecordLog,
+        observation_tx: watch::Sender<IngestV2Result<ObservationMessage>>,
+    ) -> Self {
+        let inner = InnerIngesterState {
+            shards: Default::default(),
+            rate_trackers: Default::default(),
+            replication_streams: Default::default(),
+            replication_tasks: Default::default(),
+            status: IngesterStatus::Ready,
+            observation_tx,
+        };
+        let inner = Arc::new(Mutex::new(inner));
+        let mrecordlog = Arc::new(RwLock::new(mrecordlog));
+        Self { inner, mrecordlog }
+    }
+
+    pub async fn lock_partially(&self) -> PartiallyLockedIngesterState<'_> {
+        PartiallyLockedIngesterState {
+            inner: self.inner.lock().await,
+        }
+    }
+
+    pub async fn lock_fully(&self) -> FullyLockedIngesterState<'_> {
+        FullyLockedIngesterState {
+            inner: self.inner.lock().await,
+            mrecordlog: self.mrecordlog.write().await,
+        }
```

Comment:
I suspect the order in which we acquire those locks matters a lot here, and I have actually no clue in which order they are executed in the example below.
(For instance, in C++ in initialization via constructor, it is not done in the same order as the what the code reads, but in the order of the struct members. For rust, I don't know)

I suggest the following order.
(The idea is that I suspect that the mrecordlog lock takes more time to acquire. Let me know if I am wrong)
```suggestion
        let mrecordlog_lock = self.mrecordlog.write().await;
        let inner_lock = self.inner.lock().await;
        FullyLockedIngesterState {
            inner: inner_lock,
            mrecordlog: mrecordlog_lock,
        }
```